### PR TITLE
Reinstate examples/minipyro.py, forked from pyro/examples

### PR DESCRIPTION
**Pull request by fritzo**
Joint work with @eb8680 

This replaces the outdated minipyro example with a fork of Pyro's minipyro example and adds it to our test suite.

## Tasks
- [x] update model
- [x] add unit test
- [x] fix failing tests

### CODE REVIEW

File: test/test_gaussian.py [hunk 1]
```diff
@@ -342,3 +343,15 @@ def test_integrate_gaussian(int_inputs, real_inputs):
...
+@pytest.mark.xfail(reason="numerically unstable")
+def test_mc_plate_gaussian():
+    log_measure = Gaussian(torch.tensor([0.]), torch.tensor([[1.]]),
+                           (('loc', reals()),)) + torch.tensor(-0.9189)
+    integrand = Gaussian(torch.randn((100, 1)) + 3., torch.ones((100, 1, 1)),
+                         (('data', bint(100)), ('loc', reals())))
+    with interpretation(monte_carlo):
+        res = Integrate(log_measure, integrand, frozenset({'loc'}))
+        res = res.reduce(ops.mul, frozenset({'data'}))
+        assert not torch.isinf(res).any()
```
Comment (eb8680) [inline test/test_gaussian.py]: @fritzo the minipyro example in this PR now fails with an `inf` loss.  This test demonstrates the fragment of computation within the ELBO that is the source of the numerical instability.

File: funsor/minipyro.py [hunk 2]
```diff
@@ -345,15 +345,16 @@ def step(self, *args, **kwargs):
...
     for cost in costs:
-        result += funsor.sum_product.sum_product(
-                sum_op=funsor.ops.add,
-                prod_op=funsor.ops.mul,
-                factors=probs + [cost],
-                plates=prod_vars,
-                eliminate=prod_vars | sum_vars)
+        log_prob = funsor.sum_product.sum_product(
+            sum_op=funsor.ops.logaddexp,
+            prod_op=funsor.ops.add,
```
Comment (eb8680) [inline funsor/minipyro.py]: @fritzo this fixes the numerical errors in the example, but needs more testing

File: funsor/minipyro.py [hunk 3]
```diff
@@ -345,15 +345,16 @@ def step(self, *args, **kwargs):
...
+            factors=log_probs,
+            plates=prod_vars,
+            eliminate=(prod_vars | sum_vars) - frozenset(cost.inputs)
+        )
+        result += funsor.Integrate(log_prob, cost, frozenset(cost.inputs))
```
Comment (fritzo) [inline funsor/minipyro.py]: But what if cost is plated? Or batched?

Comment (fritzo) [reply funsor/minipyro.py]: Maybe fix the `reduced_vars` arg here and add a final optional `.reduce(ops.add, ..., remaining_plates)`?

File: funsor/minipyro.py [hunk 4]
```diff
@@ -345,15 +345,18 @@ def step(self, *args, **kwargs):
...
-    probs = [p.exp() for p in log_probs]
     result = 0
     for cost in costs:
-        result += funsor.sum_product.sum_product(
-                sum_op=funsor.ops.add,
-                prod_op=funsor.ops.mul,
-                factors=probs + [cost],
-                plates=prod_vars,
-                eliminate=prod_vars | sum_vars)
+        log_prob = funsor.sum_product.sum_product(
```
Comment (eb8680) [inline funsor/minipyro.py]: In `pyro.infer.util.Dice.compute_expectation`, we use `adjoint` to share intermediate computations for costs in the same ordinal.  We could do the same thing here.
